src/esp/gfx_batch/Renderer.cpp

```diff
@@ -519,57 +667,192 @@ Cr::Containers::StridedArrayView1D<Mn::Matrix4> Renderer::transformations(
+                  scene.transformationIdsSorted.size());
+    // TODO the casting situation is GETTING OUT OF HAND
+    Mn::MeshTools::duplicateInto(
+        Cr::Containers::StridedArrayView1D<const Mn::UnsignedInt>{
+            scene.transformationIdsSorted},
+        Cr::Containers::StridedArrayView1D<
+            const Mn::Shaders::TransformationUniform3D>{
+            state_->absoluteTransformations.exceptPrefix(1)},
+        stridedArrayView(state_->absoluteTransformationsSorted.prefix(
+            scene.transformationIdsSorted.size())));
+
+    /* Upload the transformation uniforms, as they get overwritten in the
+       next loop. OTOH, interleaving it with the calculation could hide the
+       driver stalls. */
+    // TODO have this somehow in a single huge buffer instead so we can upload
+    // everything at once? ... but that would need the insane alignment
+    // requirements, not great either :(
+    scene.transformationUniform.setData(
+        state_->absoluteTransformationsSorted.prefix(
+            scene.transformationIdsSorted.size()));
+  }
 
   for (Mn::Int y = 0; y != state_->tileCount.y(); ++y) {
     for (Mn::Int x = 0; x != state_->tileCount.x(); ++x) {
       framebuffer.setViewport(Mn::Range2Di::fromSize(
           Mn::Vector2i{x, y} * state_->tileSize, state_->tileSize));
 
-      const std::size_t scene = y * state_->tileCount.x() + x;
+      const std::size_t sceneId = y * state_->tileCount.x() + x;
+      Scene& scene = state_->scenes[sceneId];
 
-      // TODO split by draw count limit
+      /* Bind buffers */
+      // TODO split by draw count limit? hard to do with those batches now, heh
+      //  also hard to do due to the insane alignment rules
       state_
           ->shader
           // TODO bind all buffers together with a multi API
           .bindProjectionBuffer(state_->projectionUniform,
-                                scene * sizeof(ProjectionPadded),
+                                sceneId * sizeof(ProjectionPadded),
                                 sizeof(ProjectionPadded))
-          .bindTransformationBuffer(state_->scenes[scene].transformationUniform)
-          .bindDrawBuffer(state_->scenes[scene].drawUniform);
+          .bindTransformationBuffer(scene.transformationUniform)
+          .bindDrawBuffer(scene.drawUniform);
       if (!(state_->flags & RendererFlag::NoTextures))
         state_->shader.bindTextureTransformationBuffer(
-            state_->scenes[scene].textureTransformationUniform);
-      state_->shader.draw(state_->mesh,
-                          stridedArrayView(state_->scenes[scene].drawCommands)
-                              .slice(&DrawCommand::indexCount),
-                          nullptr,
-                          stridedArrayView(state_->scenes[scene].drawCommands)
-                              .slice(&DrawCommand::indexOffsetInBytes));
+            scene.textureTransformationUniform);
+
+      /* Submit all draw batches */
```

Comment:
Across all scenes not really due to scissoring, otherwise objects from one scene could leak to the other. (BPS did the same IIRC.) 

But I have a bunch of ideas that could circumvent this limitation (custom clip planes, multi-view / multi-viewport rendering), just need to try them out ;)